### PR TITLE
[Datasets] [Arrow 7+ Support - 2/N] Add support for Arrow 7 by fixing Arrow serialization bug.

### LINKED ISSUE #29814: [Datasets] Arrow data buffers aren't truncated when pickling zero-copy slice views, leading to huge serialization bloat
This is a duplicating issue of https://github.com/ray-project/ray/issues/22310, so we have a clean issue on which to summarize the problem.

When pickling Arrow data (tables, in our case), Arrow currently doesn't truncate the data buffers for zero-copy slice views of the data (see [Arrow ticket](https://issues.apache.org/jira/browse/ARROW-10739)). For many of our workloads, this results is sending a potentially very large data buffer over the network for a very small slice. For pathological cases like shuffling, where we might chunk a data block into 1000 chunks and send each to a different reduce, this would involve sending 1000 copies of the entire data block; for a 1 GiB data block, we'd send an aggregate 1 TiB of data over the network instead of the expected 1 GiB. This makes Ray + Arrow essentially unusable.

This bug exists in Arrow 6 through Arrow 10; we worked around the bug in Arrow 6 via explicit copying when slicing data, e.g. in our [shuffle implementation](https://github.com/ray-project/ray/blob/095272ba0e3491a57b81a2247b24930942a261a8/python/ray/data/_internal/shuffle_and_partition.py#L59), but Arrow 7+ adds internal zero-copy slicing that makes this workaround coverage untenable (e.g. when [reading Parquet](https://github.com/ray-project/ray/issues/22310#issuecomment-1060976772)).

## Solution

We can bypass this buggy pickle path by registering a custom serializer for Arrow data that properly truncates these buffers.

### LINKED ISSUE #29992: [Datasets] Add support for using Arrow 7 with Ray (Core/Datasets/AIR).
Arrow 7 has some breaking API changes and bugs that we need to accommodate in order to make it usable with Ray.

## Known Issues

- https://github.com/ray-project/ray/issues/29814 

## Success Criteria

- CI job running against Arrow 7 is passing

### PR DESCRIPTION
**Pull request by clarkzinzow**
This PR adds support for Arrow 7 in Ray, and is the second PR in a set of stacked PRs making up this mono-PR for Arrow 7+ support: https://github.com/ray-project/ray/pull/29161, and is stacked on top of a PR fixing task cancellation in Ray Core: https://github.com/ray-project/ray/pull/29984.

## Summary of Changes

This PR:
- fixes a serialization bug in Arrow with a custom serializer for Arrow data (https://github.com/ray-project/ray/issues/29814)
- removes a bunch of defensive copying of Arrow data, which was a workaround for the aforementioned Arrow serialization bug
- adds a CI job for Arrow 7
- bumps the pyarrow upper bound to 8.0.0

## Related issue number

Closes https://github.com/ray-project/ray/issues/29992, closes https://github.com/ray-project/ray/issues/29814

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/arrow_serialization.py [hunk 1]
```diff
@@ -1,65 +1,118 @@
...
+    bug by using the Arrow IPC format for the underlying serialization.
+
+    Background:
+        Arrow has both array-level slicing and buffer-level slicing; both are zero-copy,
+        but the former has a serialization bug where the entire buffer is serialized
+        instead of just the slice, while the latter's serialization works as expected
+        and only serializes the slice of the buffer. I.e., array-level slicing doesn't
+        propagate the slice down to the buffer when serializing the array.
+
+        All that these copy methods do is, at serialization time, take the array-level
```
Comment (jianoaix) [inline python/ray/data/_internal/arrow_serialization.py:85]: This seems not relevant with the pyarrow ipc approach to handle serialization?

File: python/ray/data/datasource/parquet_datasource.py [hunk 2]
```diff
@@ -444,15 +444,19 @@ def _sample_piece(
...
-        in_memory_size = batch.nbytes / batch.num_rows
-        metadata = piece.metadata
-        total_size = 0
-        for idx in range(metadata.num_row_groups):
-            total_size += metadata.row_group(idx).total_byte_size
-        file_size = total_size / metadata.num_rows
-        ratio = in_memory_size / file_size
     except StopIteration:
         ratio = PARQUET_ENCODING_RATIO_ESTIMATE_LOWER_BOUND
+    else:
```
Comment (jianoaix) [inline python/ray/data/datasource/parquet_datasource.py:449]: Is this affected somehow or just a drive-by polish?

**Review by jianoaix — COMMENTED** [review]
Can we make PR title more specific about this fix? Or this can be done in two steps: 1) get the serialization bug fixed; 2) enable arrow 7?

File: python/ray/data/_internal/arrow_serialization.py (hunk 1 above)
Comment (clarkzinzow) [reply python/ray/data/_internal/arrow_serialization.py:85]: Ah yep this is a holdover from the old implementation, let me update this comment!

File: python/ray/data/datasource/parquet_datasource.py (hunk 2 above)
Comment (clarkzinzow) [reply python/ray/data/datasource/parquet_datasource.py:449]: Yes, we weren't handling the sampling case where the returned batch was empty, which was causing a failure here; this was a driveby fix.

**Comment by clarkzinzow** [comment]
> Or this can be done in two steps: 1) get the serialization bug fixed; 2) enable arrow 7?

This Arrow serialization bug was the core blocker for supporting Arrow 7 upgrade, and in addition to the unit testing for our serialization workaround, I think that the best validation that the serialization bug is fixed is the Arrow 7 CI job running the Datasets/AIR test suite. Since the success criteria for the serialization bug fixes is strongly tied to whether Arrow 7 works, I think that keeping them in the same PR makes sense?

File: python/ray/air/util/tensor_extensions/arrow.py [hunk 3]
```diff
@@ -470,12 +470,7 @@ def __getitem__(self, key):
         # unfortunately overriding Cython cdef methods with normal Python methods isn't
         # allowed.
         if isinstance(key, slice):
-            sliced = super().__getitem__(key).to_numpy()
-            if sliced.dtype.type is not np.object_:
-                # Force ths slice to match NumPy semantics for unit (single-element)
-                # slices.
-                sliced = sliced[0:1]
-            return sliced
+            return super().__getitem__(key)
```
Comment (clarkzinzow) [inline python/ray/air/util/tensor_extensions/arrow.py:473]: The Arrow extension array contract is generally to another extension array for a slice, and the element type (in our case an ndarray) when getting a single row.

File: python/ray/data/datasource/parquet_datasource.py [hunk 4]
```diff
@@ -434,7 +434,9 @@ def _sample_piece(
 
     # Only sample the first row group.
     piece = piece.subset(row_group_ids=[0])
-    batch_size = min(piece.metadata.num_rows, PARQUET_ENCODING_RATIO_ESTIMATE_NUM_ROWS)
+    batch_size = max(
+        min(piece.metadata.num_rows, PARQUET_ENCODING_RATIO_ESTIMATE_NUM_ROWS), 1
```
Comment (clarkzinzow) [inline python/ray/data/datasource/parquet_datasource.py:438]: Arrow doesn't allow a batch size of 0; this is needed to handle empty Parquet files.

File: python/ray/data/_internal/arrow_block.py [hunk 5]
```diff
@@ -70,6 +71,19 @@ class ArrowRow(TableRow):
...
 
     def __getitem__(self, key: str) -> Any:
+        from ray.data.extensions.tensor_extension import (
+            ArrowTensorType,
+            ArrowVariableShapedTensorType,
+        )
+
+        schema = self._row.schema
+        if isinstance(
+            schema.field(schema.get_field_index(key)).type,
```
Comment (c21) [inline python/ray/data/_internal/arrow_block.py:81]: nit: can it just be `schema.field(key).type` - https://arrow.apache.org/docs/python/generated/pyarrow.Schema.html#pyarrow.Schema.field ?

File: python/ray/data/_internal/arrow_serialization.py [hunk 6]
```diff
@@ -1,65 +1,118 @@
...
+
+        We work around this by registering a custom cloudpickle reducers for Arrow
+        Tables that delegates serialization to the Arrow IPC format; thankfully, Arrow's
+        IPC serialization has fixed this buffer truncation bug.
+
+    See https://issues.apache.org/jira/browse/ARROW-10739.
+    """
+    import pyarrow as pa
+
+    if os.environ.get(RAY_DISABLE_CUSTOM_ARROW_DATA_SERIALIZATION, "0") == "1":
```
Comment (c21) [inline python/ray/data/_internal/arrow_serialization.py:93]: Glad that we add an environment variable here!

**Review by c21 — APPROVED** [review]
LGTM with some minor questions/comments.

**Review by jianoaix — APPROVED** [review]
Look good, thanks for making it simpler with Arrow IPC

File: python/ray/data/datasource/parquet_datasource.py (hunk 4 above)
Comment (clarkzinzow) [reply python/ray/data/datasource/parquet_datasource.py:438]: Yep filtering out empty files when sampling is a good idea! It should be easy enough to filter out empty files before constructing the sample space: https://github.com/ray-project/ray/blob/abc0bb0a625e01aee9f2ccd777248892db2a5f6c/python/ray/data/datasource/parquet_datasource.py#L303-L308

File: python/ray/data/_internal/arrow_block.py (hunk 5 above)
Comment (clarkzinzow) [reply python/ray/data/_internal/arrow_block.py:81]: Hmm I had thought that this support was added in a post-Arrow 6 version, but looks like it's supported. I'll make this change!

File: python/ray/data/_internal/delegating_block_builder.py [hunk 7]
```diff
@@ -59,6 +59,7 @@ def build(self) -> Block:
         if self._builder is None:
             if self._empty_block is not None:
                 self._builder = BlockAccessor.for_block(self._empty_block).builder()
+                self._builder.add_block(self._empty_block)
```
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply python/ray/data/_internal/delegating_block_builder.py:62]: Yes, I believe that this was also for the empty Parquet file case, and should be covered by the added test!

**Comment by clarkzinzow** [comment]
@ericl @scv119 @jjyao @richardliaw Could I get a rubber stamp for the `setup.py` changes?

**Review by scv119 — APPROVED** [review]
image

**Comment by clarkzinzow** [comment]
Ah sorry @scv119 I thought that you had the magic powers, but looks like it needs to be one of @ericl @richardliaw or @edoakes!

**Review by ericl — APPROVED** [review]
dep spec LGTM
